doc/design: document postgres sources design

petrosagg opened the pull request:
[**View rendered.**](https://github.com/petrosagg/materialize/blob/6de21488ce391b709a89af111fbd5398a52df711/doc/developer/design/20210412_postgres_sources.md)

Many users are keen to use materialize with a dataset existing in a postgres
upstream database. The current best-practice to ingest the data in realtime is
to setup a Kafka cluster and use Debezium to convert the postgres CDC format
into Kafka topics that can then be consumed by materialize.

This design doc proposes to add a new type of source called `POSTGRES` where
materialized directly connects to an upstream database and interprets the
replication protocol, ingesting row updates in real time.

Users should be able to try materialize out simply by running the materialize
binary and pointing it to their postgres database, no extra infrastructure
needed.

mjibson commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+  stream even when `REPLICA IDENTITY` is set to `FULL`. If a TOASTed value is
+  encountered the source will enter an errored state.
+
+* **The TCP connection to the upstream database will not drop**. Resuming a
```
What happens if mz needs to restart?

mjibson commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+BEGIN message. This allows the source implementation to emit rows as it receives
+them and ensure that transaction boundaries are guaranteed.
+
+### Adaptation to multiple tables
```
I think you also need to add discussion about how SimpleSource will be extended to support multiple tables?

petrosagg replied:
I'm not really sure how this will be done yet and also there is still a possibility of using the `SourceInfo` trait instead if it proves to be easier. I can add this discussion once the implementation has progressed a bit

mjibson commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+which is what we're trying to get to in the first place.
+
+There exist subcases of this problem that make it easier. For example, upstream
+tables with an auto-increment primary key have a natural way of resuming form
```
This isn't quite true. Auto-increment is only guaranteed to be atomic, but not transactional. Auto-increment PKs can be inserted out of order and there can be skipped values.

petrosagg replied:
Ah, yeah that's a good point. So we're back to "reconnect during initial sync is hard" territory

petrosagg replied:
Removed this section

mjibson commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+tables with an auto-increment primary key have a natural way of resuming form
+where we left off, just remember the last id. Finally, a naive implementation
+could buffer the initial snapshot in-memory, outside of the dafalow, and only
+issue the data into the dataflow once we've received a full snapshot.
```
How would this work exactly? Once we receive the full snapshot, we read all of the dataflow and retract each point, and then insert all the points from the snapshot?

petrosagg replied:
No, it would be the other way around. In the initial sync we buffer everything in memory in the source and don't issue anything in the dataflow.

If the connection is interrupted we can just throw away our buffered data and start from the beginning.

Once we receive a full snapshot we can issue data in the dataflow and not worry about disconnections again because even if they happen we now have the replication slot that supports resumption

mjibson reviewed (commented):
LGTM, with the caveat that I think the resume story should be explored more to make sure that there's a reasonable way to resume.

umanwizard commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+  currently supports enough types to cover most use cases. If an upstream tables
+  uses an unsupported type source creation will not be allowed.
+
+* **All tables will have `REPLICA IDENTITY` set to `FULL`.** This will force
```
This can be avoided with some `UPSERT`-like logic, if the memory overhead is acceptable

umanwizard commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+* Support a suite of common column data types that cover most use cases
+* Materialize all synced tables by default. That is, all data must fit in RAM
+
+## Non-Goals
```
I understand the desire to limit scope, but it would be useful for each non-goal to say briefly whether we have a plan and are working towards it, or if it hasn't been investigated at all.

petrosagg replied:
Added a couple of sentences for each non-goal to that effect

cuongdo commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+  detected, materialize will put the relevant source in an errored state and
+  wait for an administrator to fix the problem.
+
+* **All synced tables will be materialized.** Normally, sources in materialized
```
This limits the usefulness of PG replication relative to PG/Debezium/Kafka sources, because the latter allows you to materialize only a small subset of the source data. Is it feasible to loosen this requirement after MVP?

cuongdo commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+* **The upstream database doesn't require complex authentication or TLS
+  connections.** After the MVP, simple TLS connections should be prioritized.
+
+* **The schema will remain static for the duration of the source.** When
```
How does this compare to our schema evolution story for Debezium sources?

petrosagg replied:
I'll read up on our strategy with debezium, I'm not sure how we handle it there

umanwizard replied:
I can answer this for Debezium. When a table changes in a compatible way (e.g. adding columns) we will continue to ingest data according to the old schema (e.g. ignoring the new added column).

This mostly works but it has been the source of bugs in the past (the customer issues we had last week were due to this happening).

cuongdo commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+* **All synced tables will be materialized.** Normally, sources in materialized
+  can be instantiated on demand and each instance talks to external systems
+  independently. This is a bad pattern for an upstream postgres database because
+  each instance would have to create a replication slot, which is a scarce
```
If they're not cleaned up, replication slots can cause PG WAL files to stay around. This can result in the upstream database running out of disk space and crashing. I think it is critical to test this scenario.

petrosagg replied:
Yeah that's right. Right now we're using temporary replication slots which are tied to the current pg session and cleaned up automatically on disconnect but when we implement resumption we'll have to keep them around and clean them up on source DROP on our side.

cuongdo commented on doc/developer/design/20210412_postgres_sources.md:
````diff
@@ -0,0 +1,210 @@
+
+```sql
+-- Initial sync phase
+BEGIN READ ONLY ISOLATION LEVEL REPEATABLE READ
````
What impact does this have on the database performance? Can this be run on a primary database, or should we recommend that people replicate tables from a replica database?

umanwizard replied:
Yes, but you would only have to provision the Pg read replica once, and then if you are bringing Materialize up and down multiple times (or it crashes) that load would go to the replica.

petrosagg replied:
That's a good point. So I think our documentation should say that every time materialize starts up it's as if a new postgres replica is connected to the upstream database, and let the db administrators decide if this an acceptable load for their primary

cuongdo commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+with other sources in materialize. This means that all updates must have a
+timestamp that is related to the wall-clock time.
+
+The implementation is greatly simplified by the logical replication protocol
```
Do we have to do anything special to handle nested transactions (nested BEGINs/ENDs)?

cuongdo commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+There exist subcases of this problem that make it easier. For example, upstream
+tables with an auto-increment primary key have a natural way of resuming form
+where we left off, just remember the last id. Finally, a naive implementation
+could buffer the initial snapshot in-memory, outside of the dafalow, and only
```
nit: dafalow -> dataflow

cuongdo reviewed (commented):
Looks good overall. This should also have a testing strategy, either as part of the doc or as a separate GitHub issue. Failures can occur during various parts of our network interactions with Postgres, and we should be prepared to simulate those.

cuongdo commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+* Sync the initial state of the database and seamlessly switch to streaming
+* Guarantee that transaction boundaries across tables will be preserved
+* Support a suite of common column data types that cover most use cases
+* Materialize all synced tables by default. That is, all data must fit in RAM
```
This is fine for MVP, but I'm concerned if this is a long-term constraint. For some databases, on-disk tables can be much larger than fits comfortably in memory.

umanwizard replied:
I 100% agree that supporting unmaterialized Pg sources is a crucial feature. We've discussed this on the team and I think there is consensus on that point.

petrosagg replied:
I agree. In order to support that we need to implement the dynamic resource sharing proposal that is quite a bit of work.

However I don't think this should block the release of the feature (maybe it should block version 1.0 specifically) because there are many usecases where the data fits comfortably in RAM.

frankmcsherry replied:
I think this means moving "unmaterialized sources" to a non-goal? At least, if we all agree that this is a temporary limitation rather than an intent of postgres sources.

ruchirK commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+
+A key design goal is to maintain the  transaction boundaries from the upstream
+database so that no query on materialize can observe states that never existed
+upstream. At the same time, the data ingested from postgres should be join-able
```
I think perhaps for the mvp we could relax the "should be join-able with other sources" requirement and instead just use the lsn as the timestamp / some other field that easily generalizes to multiple tables?

ruchirK reviewed (approved):
I mostly focused on the timestamps portion here. I think everything so far makes sense and is maybe waiting on some more clarity for "how are we going to assign / persist real time timestamps for postgres sources" (probably waiting on me honestly :D ) -- I think I would say maybe the first MVP should just do non-real time timestamps so that we can get something less complicated out?

JLDLaughlin commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+
+By creating the replication slot as the first statement in a transaction and
+setting the `USE_SNAPSHOT` option we ensure that the transaction and the slot
+are at the exact same point. The existence of the slot will prevent any WAL
```
> The existence of the slot will prevent any WAL segments from being cleaned up...

Since the slot will outlive the `COPY` process, does this mean that WAL segments will not be able to be cleaned up for the entirety of the replication process? (As long as a Postgres source in Materialize is live?) Or does something about the slot change when the `COPY` step has completed?

JLDLaughlin replied:
I think I answered my own question from [the Postgres docs!](https://www.postgresql.org/docs/13/warm-standby.html#STREAMING-REPLICATION-SLOTS) 
> Replication slots provide an automated way to ensure that the master does not remove WAL segments until they have been received by all standbys, and that the master does not remove rows which could cause a recovery conflict even when the standby is disconnected.

petrosagg replied:
The specific mechanism is through the standby status update messages in the protocol that let the upstream know up to what LSN is safe to clean up

JLDLaughlin commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+### Adaptation to multiple tables
+
+If the sources naively applied the mechanism described above for every source
+instance they would very quickly exhaust the available replication slots from
```
It might be worth including a link to the [documented replication limits](https://www.postgresql.org/docs/13/runtime-config-replication.html), ex: `max_replication_slots` is 10 by default.

petrosagg replied:
added!

JLDLaughlin commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+important one as this only happens on source startup. In order to handle
+disconnections during this phase the implementation needs to be able to issue a
+brand new COPY command at the same point as the interrupted transfer. Normally,
+a system could just flush its in-flight state and start anew, but since we're
```
Would it be unreasonable to drop and re-ship a dataflow instead of issuing retractions?

petrosagg replied:
Can this be done safely? What if the data has reached some external sink?

petrosagg commented:
@cuongdo added a section for the testing strategy.
@ruchirK if it turns out that assigning realtime timestamps is hard I will fall back to using LSNs

I also removed the section and assumption of having a stable TCP connection as resumption is now implemented! https://github.com/MaterializeInc/materialize/pull/6477

frankmcsherry commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,210 @@
+database so that no query on materialize can observe states that never existed
+upstream. At the same time, the data ingested from postgres should be join-able
+with other sources in materialize. This means that all updates must have a
+timestamp that is related to the wall-clock time.
```
Does it mean this? Many SQL systems have *system versioned time* and if they present that outwards it seems like it would be a very reasonable timestamp to use.

petrosagg replied:
What I mean is that currently in materialize there are BYO sources that have a timestamp value that could represent anything and realtime sources that are based on wall clock of ingestion.

You're right that two BYO sources could be using compatible system times and be perfectly join-able together. I'll re-write this sentence to make more clear

frankmcsherry replied:
Maybe a different framing from me then: It would be best if we mirrored the upstream systems view of times as faithfully as possible, and then in Materialize determined if we wanted to convert timestamps to local "materialize time", rather than baking in materialize time to the definition of the sources.

frankmcsherry commented on doc/developer/design/20210412_postgres_sources.md:
```diff
@@ -0,0 +1,214 @@
+have one `SOURCE` object created for each upstream table. By introducing this
+syntax the SQL planner will have an opportunity to create all the relevant
+sources at once and therefore give them some shared state. The shared state
+could be some form of `Arc<RwLock<Option<T>>>` where the first source instance
```
This came up in discussion. This is potentially problematic if it is meant to be shared across workers, but not problematic if it is to be shared across sources but within a worker. I.e., if each of the sources behave a bit like we might want Kafka sources to, sharing the reader for multiple source instances, the no worries and we could mirror that design. If instead the goal is to get some coordination between the workers using `Arc<Mutex<_>>` we should strongly avoid that imo (and instead try and use dataflow to coordinate, if at all possible). One example from the chat was to have one operator that reads single-threaded from postgres (idk) and then distributes its readings to independent operators downstream for each table; this "coordinates" the reading from postgres but still gets the results to distributed source instances.

petrosagg replied:
Hm, in the case of postgres sources specifically it is true that the reading of the replication stream is single threaded and also that all the downstream sources are completely dumb and just report the messages received in their collections.

Does what you say mean that we could force all of these sources to run on the same worker and use an `Arc<Mutex<..>>` to spawn the reading task once?

cuongdo commented:
@petrosagg & @JLDLaughlin what work remains before this design doc can be merged?

JLDLaughlin commented:
@cuongdo this actually needs to be updated to reflect conversations that happened last week! Once it's updated, I think we're good to merge.
